platforms/ide/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/TestOperationMapper.java

```diff
@@ -84,12 +86,46 @@ public InternalOperationFinishedProgressEvent createFinishedEvent(DefaultTestDes
-        return new DefaultTestDescriptor(buildOperationId, suite.getName(), suite.toString(), suite.getDisplayName(), InternalJvmTestDescriptor.KIND_SUITE, suite.getName(), suite.getClassName(), suite.getMethodName(), parentId, taskTracker.getTaskPath(buildOperationId));
+    private DefaultTestDescriptor toTestDescriptorForSuite(OperationIdentifier buildOperationId, OperationIdentifier parentId, TestDescriptor suite) {
+        String methodName = null;
+        String operationDisplayName = suite.toString();
+
+        TestDescriptor originalDescriptor = getOriginalDescriptor(suite);
+        if (originalDescriptor instanceof AbstractTestDescriptor) {
+            methodName = ((AbstractTestDescriptor) originalDescriptor).getMethodName();
+        } else {
+            operationDisplayName = getLegacyOperationDisplayName(operationDisplayName, originalDescriptor);
+        }
+        return new DefaultTestDescriptor(buildOperationId, suite.getName(), operationDisplayName, suite.getDisplayName(), InternalJvmTestDescriptor.KIND_SUITE, suite.getName(), suite.getClassName(), methodName, parentId, taskTracker.getTaskPath(buildOperationId));
+    }
+
+    private DefaultTestDescriptor toTestDescriptorForTest(OperationIdentifier buildOperationId, OperationIdentifier parentId, TestDescriptor test) {
+        String operationDisplayName = test.toString();
+
+        TestDescriptor originalDescriptor = getOriginalDescriptor(test);
+        if (!(originalDescriptor instanceof AbstractTestDescriptor)) {
+            operationDisplayName = getLegacyOperationDisplayName(operationDisplayName, originalDescriptor);
+        }
+        return new DefaultTestDescriptor(buildOperationId, test.getName(), operationDisplayName, test.getDisplayName(), InternalJvmTestDescriptor.KIND_ATOMIC, null, test.getClassName(), test.getName(), parentId, taskTracker.getTaskPath(buildOperationId));
     }
 
-    private DefaultTestDescriptor toTestDescriptorForTest(OperationIdentifier buildOperationId, OperationIdentifier parentId, TestDescriptorInternal test) {
-        return new DefaultTestDescriptor(buildOperationId, test.getName(), test.toString(), test.getDisplayName(), InternalJvmTestDescriptor.KIND_ATOMIC, null, test.getClassName(), test.getName(), parentId, taskTracker.getTaskPath(buildOperationId));
+    // workaround for some plugins that override TestDescriptor
+    private static String getLegacyOperationDisplayName(String operationDisplayName, TestDescriptor testDescriptor) {
+        // if toString() is not overridden, use the display name for test operation
+        if (operationDisplayName.endsWith("@" + Integer.toHexString(testDescriptor.hashCode()))) {
+            return testDescriptor.getDisplayName();
+        } else {
+            return operationDisplayName;
+        }
+    }
+
+    // can be removed once the workaround above is removed
```

Comment:
❗This comment is too loose on details: How can we know that the workaround is no more, so we could actually remove this? Also, is there an issue which would trigger this fix?

Reply:
+1